Lib/asyncio/tasks.py

```diff
@@ -408,19 +408,32 @@ async def wait(fs, *, timeout=None, return_when=ALL_COMPLETED):
     when the timeout occurs are returned in the second set.
     """
     if futures.isfuture(fs) or coroutines.iscoroutine(fs):
-        raise TypeError(f"expect a list of futures, not {type(fs).__name__}")
-    if not fs:
-        raise ValueError('Set of Tasks/Futures is empty.')
+        raise TypeError(f"expected an iterable of futures, not {type(fs).__name__}")
     if return_when not in (FIRST_COMPLETED, FIRST_EXCEPTION, ALL_COMPLETED):
         raise ValueError(f'Invalid return_when value: {return_when}')
 
-    fs = set(fs)
+    loop = events.get_running_loop()
+    new_fs = set()
 
-    if any(coroutines.iscoroutine(f) for f in fs):
-        raise TypeError("Passing coroutines is forbidden, use tasks explicitly.")
+    for f in fs:
+        if coroutines.iscoroutine(f):
+            raise TypeError("Passing coroutines is forbidden, use tasks explicitly.")
+        if not futures.isfuture(f):
+            warnings.warn(
```

Comment:
Use `warnings._deprecated`?

Reply:
Yeah I usually would, but I kinda want to try backport the deprecation fix to 3.10

also the wording `warnings._deprecated` doesn't work to well for changes in behaviour and I didn't want to try and fix it and open a new can of worms there

Reply:
@hugovk are you already working on a plan to upbraid all the existing  `warnings.warn(..., DeprecationWarning)` to `warnings._deprecated` ?

there's a slight usability/wording concern when using it for changes in behavior instead of actual full removals
and there also seems to be a desire to use it for totally unscheduled removals: `remove=(4, )`

<hr>

in addition do you have a view on backporting deprecation, where it was clear from the code what the intention of the deprecation was, but the deprecation itself had a bug?

Reply:
> @hugovk are you already working on a plan to upbraid all the existing `warnings.warn(..., DeprecationWarning)` to `warnings._deprecated` ?

I'm not but that's a good idea. From a quick check, there's not too many, and most are set to be removed in 3.12 anyway.

> there's a slight usability/wording concern when using it for changes in behavior instead of actual full removals 

Can you alter the wording using the `message` parameter?

> and there also seems to be a desire to use it for totally unscheduled removals: `remove=(4, )`

Some removals scheduled for 4.0 were from before we knew if 3.10 or 4.0 would follow 3.9 (https://github.com/python/cpython/issues/80480#issuecomment-1093818166) and in at least one case is shorthand for "sometime after 2.7.final" (https://github.com/python/cpython/issues/80480#issuecomment-1093818167).

Anyway, we should review any remaining removals originally set for 4.0 (e.g. in new issues or relevant open issues) to decide if they can be removed in an upcoming 3.x.

---

> in addition do you have a view on backporting deprecation, where it was clear from the code what the intention of the deprecation was, but the deprecation itself had a bug?

Generally, I think it's a good idea to increase visibility and help users upgrade in time to reduce friction. It's good to ask release managers when in doubt.

Reply:
@hugovk do you think we have enough here to make a new issue out of this side discussion. Or maybe there's two to four ideas here:

* Do something about existing deprecation warnings that don't automatically get upgraded to errors
* Do something about the wording and format-string templating in warnings._deprecated to make it work for even the most subtle behaviour change
* Determine a backport strategy for when a deprecation warning was applied to the wrong thing
* Determine a strategy for permanent deprecations, where a feature is a bit broken or an alias of another thing, and people should really use the not broken way or source of the alias, but CPython is unwilling or undecided on actually removing it

Reply:
Sure! Or at https://discuss.python.org/?